src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java

```diff
@@ -118,102 +152,200 @@ private AuthorizationFramework() {
+                        getClassName(plugin));
+                unloadPlugin(wrapper);
+            }
+            wrapper.setPlugin(plugin);
+            return;
         }
+        LOGGER.log(Level.INFO, "Plugin class \"{0}\" was not found in configuration."
+                + " Appending the plugin at the end of the list with role \"{1}\"",
+                new Object[]{getClassName(plugin), role});
+        plugins.add(new AuthorizationPluginWrapper(role, getClassName(plugin), plugin));
     }
 
-    private void removeAll() {
-        synchronized (this) {
-            for (IAuthorizationPlugin plugin : getPlugins()) {
-                try {
-                    plugin.unload();
-                } catch (Throwable ex) {
-                    LOGGER.log(Level.SEVERE, "Plugin \"" + plugin.getClass().getName() + "\" has failed while unloading with exception:", ex);
-                }
-            }
-            plugins.clear();
+    /**
+     * Remove and unload a single plugin.
+     *
+     * @param plugin the plugin wrapper
+     * @see AuthorizationPluginWrapper#unload()
+     */
+    protected synchronized void removePlugin(AuthorizationPluginWrapper plugin) {
+        unloadPlugin(plugin);
+        plugins.remove(plugin);
+    }
+
+    /**
+     * Remove and unload all plugins.
+     *
+     * @see AuthorizationPluginWrapper#unload()
+     */
+    public synchronized void removeAll() {
+        unloadAllPlugins();
+        plugins = new ArrayList<>();
+    }
+
+    /**
+     * Call unload and load method on the plugin.
+     *
+     * @param plugin the plugin wrapper
+     * @see AuthorizationPluginWrapper#unload()
+     * @see AuthorizationPluginWrapper#load()
+     */
+    protected synchronized void reloadPlugin(AuthorizationPluginWrapper plugin) {
+        unloadPlugin(plugin);
+        loadPlugin(plugin);
+    }
+
+    /**
+     * Call the load method on the plugin and set it as working if it does not
+     * fail.
+     *
+     * @param plugin the plugin wrapper
+     * @see AuthorizationPluginWrapper#load()
+     */
+    protected synchronized void loadPlugin(AuthorizationPluginWrapper plugin) {
+        try {
+            plugin.load();
+            plugin.setWorking();
+        } catch (Throwable ex) {
+            LOGGER.log(Level.SEVERE, "Plugin \"" + plugin.getClassname() + "\" has failed while loading with exception:", ex);
+            plugin.setFailed();
         }
     }
 
     /**
-     * @param suffix suffix for the files
-     * @return list of file with suffix
+     * Load all plugins in the current array. If the plugin has not been loaded
+     * yet it is marked as failed.
      */
-    private List<File> listFiles(String suffix) {
-        File[] files = directory.listFiles(new FilenameFilter() {
-            @Override
-            public boolean accept(File dir, String name) {
-                return name.endsWith(suffix);
+    public void loadAllPlugins() {
```

Comment:
it would be good to log the list of plugins (including their order) at the end of the (re)loading so that we know how the stack looks like.